File: backend/app/api/v1/endpoints/company_insights.py
"""
Lead Intelligence & Company Analysis endpoints (Module 4).

Nested under a lead (`/leads/{lead_id}/insights`), since an insight only
ever makes sense in the context of one lead — there is no standalone
"list all insights" use case in the product.
"""

import uuid

from fastapi import APIRouter, status

from app.api.deps import AIProviderDep, CurrentActiveUser, DBSession
from app.core.exceptions import NotFoundError
from app.schemas.company_insight import CompanyInsightRead
from app.services.company_intelligence_service import CompanyIntelligenceService

router = APIRouter()


@router.post(
    "/{lead_id}/insights/generate",
    response_model=CompanyInsightRead,
    status_code=status.HTTP_201_CREATED,
    summary="Generate a new AI company-insight snapshot for a lead",
)
async def generate_insight(
    lead_id: uuid.UUID, db: DBSession, current_user: CurrentActiveUser, ai_provider: AIProviderDep
) -> CompanyInsightRead:
    insight = await CompanyIntelligenceService(db, ai_provider).generate_insight(lead_id, current_user)
    return CompanyInsightRead.model_validate(insight)


@router.get(
    "/{lead_id}/insights/latest",
    response_model=CompanyInsightRead,
    summary="Get the most recent company insight for a lead",
)
async def get_latest_insight(
    lead_id: uuid.UUID, db: DBSession, current_user: CurrentActiveUser, ai_provider: AIProviderDep
) -> CompanyInsightRead:
    insight = await CompanyIntelligenceService(db, ai_provider).get_latest_insight(lead_id, current_user)
    if insight is None:
        raise NotFoundError(
            "No insight has been generated for this lead yet.", error_code="insight_not_found"
        )
    return CompanyInsightRead.model_validate(insight)


@router.get(
    "/{lead_id}/insights",
    response_model=list[CompanyInsightRead],
    summary="List all company-insight snapshots for a lead",
)
async def list_insights(
    lead_id: uuid.UUID, db: DBSession, current_user: CurrentActiveUser, ai_provider: AIProviderDep
) -> list[CompanyInsightRead]:
    insights = await CompanyIntelligenceService(db, ai_provider).list_insights(lead_id, current_user)
    return [CompanyInsightRead.model_validate(insight) for insight in insights]

File: frontend/src/services/api/companyInsights.js
import api from "./client";

// Lead Intelligence & Company Analysis (Module 4 on the backend).
// Every insight is scoped to a lead, hence the nested /leads/{leadId}/insights path.

// Generate a brand-new AI company-insight snapshot for a lead.
export const generateCompanyInsight = async (leadId) => {
  const response = await api.post(`/leads/${leadId}/insights/generate`);
  return response.data;
};

// Get the most recently generated insight for a lead.
// Returns null (instead of throwing) when no insight has been generated yet,
// since a 404 here is an expected, normal state for a fresh lead.
export const getLatestCompanyInsight = async (leadId) => {
  try {
    const response = await api.get(`/leads/${leadId}/insights/latest`);
    return response.data;
  } catch (error) {
    if (error?.response?.status === 404) {
      return null;
    }
    throw error;
  }
};

// List every insight snapshot ever generated for a lead (most recent history).
export const listCompanyInsights = async (leadId) => {
  const response = await api.get(`/leads/${leadId}/insights`);
  return response.data;
};
